src/org/mozilla/javascript/SpecialRef.java

```diff
@@ -105,11 +104,18 @@ public Object set(Context cx, Scriptable scope, Object value)
                     }
 
                     if (cx.getLanguageVersion() >= Context.VERSION_ES6) {
-                        if ((value != null && !"object".equals(ScriptRuntime.typeof(value))) ||
-                            !"object".equals(ScriptRuntime.typeof(target))) {
-                            return Undefined.instance;
+                        final String typeOfValue = ScriptRuntime.typeof(value);
+                        if (value instanceof Number) {
+                            return value;
+                        }
+                        if (!"object".equals(typeOfValue) && !"function".equals(typeOfValue)) {
+                            return obj;
```

Comment:
Probably for non-objects, this should always be `value` rather than `obj`:
```suggestion
                        if (!"object".equals(typeOfValue) && !"function".equals(typeOfValue)) {
                            return value;
```

I tested using this code:

```js
[undefined, null, true, 123, "abc", Symbol(), function () {}, {}].forEach(value => {
    var obj = {}
    var res = (obj.__proto__ = value);
    console.log(typeof value + ": returned=" + (res === value))
})
```